Reimplement LogManager sysmodule

XorTroll opened the pull request:
Reimplements LogManager (LogManager.Prod in retail) sysmodule, redirecting log data to log files in the SD card.
There isn't much left to do (everything seems to work fine), it just needs testing and a proper review.

XorTroll commented:
So, one of the aspects I have in my mind right now is the idea of specifying log verbosity via an atmosphere setting, since these logs have many fields which could be logged, but logging all of them (what this reimplementation currently does) might be too verbose for users/devs which just want the text log and don't care about the rest of the fields.

misson20000 commented:
I'd rather see logs stores in a machine-readable format for later searching and filtering.

XorTroll commented:
This last commit changed many things:
- Now logs are saved in binary format (agree, it's better to save them this way and then parse/print them elsewhere). Save format: `sdmc:/atmosphere/debug_logs/<hex-program-id>/<hex-log-id>.bin`, where the log ID is the current system tick.
- Logs are removed everytime the sysmodule starts (every reboot), since games like Rocket League generate tons of logs which end up wasting a lot of space (in this case, RL generates hundreds of logs just before the title screen is loaded).
- As a way to let other sysmodules/processes get advantage of our logging, added two extra commands to lm service: one to get an event which signals when any process logs, and another one to get the last log's information (log ID and program ID), so that the sysmodule can read the logged data from SD.
- Removed all log packet decoding code, since it is no longer used in LogManager.
As a sidenote, I've tested the extra commands with a simple TCP logger between PC and a sysmodule, and I managed to print everything games/processes log in real time on PC without any issues.

SciresM commented on stratosphere/LogManager/LogManager.json:
```diff
@@ -0,0 +1,103 @@
+    "title_id_range_min": "0x0100000000000015",
+    "title_id_range_max": "0x0100000000000015",
+    "main_thread_stack_size": "0x00004000",
+    "main_thread_priority": 10,
```
This is incorrect. LogManager.MainThread has a system thread priority of 10; this corresponds to an actual hos priority of 38.

SciresM commented on stratosphere/LogManager/LogManager.json:
```diff
@@ -0,0 +1,103 @@
+            "value": {
+                "allow_debug": false,
+                "force_debug": true
+            }
```
Should there be a handle table descriptor to match official LogManager?

SciresM commented on stratosphere/LogManager/source/impl/lm_logging.cpp:
```diff
@@ -0,0 +1,92 @@
+#include "lm_logging.hpp"
```
Atmosphere copyright header + #include <stratosphere.hpp>

SciresM commented on stratosphere/LogManager/source/impl/lm_logging.cpp:
```diff
@@ -0,0 +1,92 @@
+        g_can_access_fs = can_access;
+    }
+
+    Result WriteLogPackets(std::vector<LogPacketBuffer> &packet_list) {
```
std::vector --> no.

SciresM commented on stratosphere/LogManager/source/lm_service.cpp:
```diff
@@ -0,0 +1,77 @@
+
+    }
+
+    void Logger::Log(const sf::InAutoSelectBuffer &buf) {
```
This doesn't seem to match the code RE I've done on lm in the past...

SciresM commented on stratosphere/LogManager/source/lm_types.hpp:
```diff
@@ -0,0 +1,14 @@
+namespace ams::lm {
+
+    enum class LogDestination : u32 {
+        TMA = 1,
```
Atmosphere uses (1 << ...) for bitflag enums.

SciresM commented on stratosphere/Makefile:
```diff
@@ -1,4 +1,4 @@
-MODULES := loader ncm pm sm boot ams_mitm spl eclct.stub ro creport fatal dmnt boot2 erpt pgl jpegdec
+MODULES := loader ncm pm sm boot ams_mitm spl eclct.stub ro creport fatal dmnt boot2 erpt pgl jpegdec LogManager
```
module should really be called lm (and just have LogManager as npdm name)

SciresM commented on stratosphere/LogManager/source/lm_service.hpp:
```diff
@@ -0,0 +1,59 @@
+
```
Atmosphere copyright header, #include <stratosphere.hpp>

SciresM commented on stratosphere/LogManager/source/lm_service.cpp:
```diff
@@ -0,0 +1,77 @@
+#include "lm_service.hpp"
```
Atmosphere copyright header, #include <stratosphere.hpp>

SciresM commented on stratosphere/LogManager/source/lm_main.cpp:
```diff
@@ -0,0 +1,150 @@
+#include "lm_service.hpp"
```
Atmosphere copyright header, #include <stratosphere.hpp>

SciresM commented on stratosphere/LogManager/source/impl/lm_logging.hpp:
```diff
@@ -0,0 +1,112 @@
+
```
Atmosphere copyright header, #include <stratosphere.hpp>

SciresM commented on stratosphere/LogManager/source/lm_types.hpp:
```diff
@@ -0,0 +1,14 @@
+
```
Atmosphere copyright header.

SciresM commented on stratosphere/LogManager/source/lm_service.cpp:
```diff
@@ -0,0 +1,77 @@
+    void LogService::OpenLogger(const sf::ClientProcessId &client_pid, sf::Out<std::shared_ptr<Logger>> out_logger) {
+        u64 program_id = 0;
+        /* Apparently lm succeeds on many/all commands, so we will succeed on them too. */
+        pminfoGetProgramId(&program_id, static_cast<u64>(client_pid.GetValue()));
```
This should use the libstratosphere  wrapper.

SciresM commented on stratosphere/LogManager/source/lm_main.cpp:
```diff
@@ -0,0 +1,150 @@
+
+namespace {
+
+    /* TODO: these domain/domain object amounts work fine, but which ones does N actually use? */
```
Yeah, you should check what N uses.

SciresM commented on libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_packet_transmitter.hpp:
```diff
@@ -0,0 +1,122 @@
+        NON_COPYABLE(LogPacketTransmitterBase);
+        NON_MOVEABLE(LogPacketTransmitterBase);
+        public:
+            using FlushFunction = bool (*)(const void*, size_t);
```
FlushFunction takes in const u8 *, not const void *.

SciresM commented on libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_packet_transmitter.hpp:
```diff
@@ -0,0 +1,122 @@
+            using FlushFunction = bool (*)(const void*, size_t);
+        private:
+            LogPacketHeader *header;
+            u8 *log_buffer_start;
```
Kind of verbose, consider e.g. *start; *end; *payload; *current;

This is not a required change.

SciresM commented on libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_packet_transmitter.hpp:
```diff
@@ -0,0 +1,122 @@
+            bool is_tail;
+            FlushFunction flush_function;
+        public:
+            LogPacketTransmitterBase(u8 *log_buffer, size_t log_buffer_size, FlushFunction flush_fn, diag::LogSeverity severity, bool verbosity, u64 process_id, bool head, bool tail);
```
log_buffer should be void *, do not abbreviate function as fn. use "func", if you must..

Severity/verbosity are u8, not enum/bool here.

SciresM commented on libraries/libstratosphere/include/stratosphere/fs/fs_sd_card.hpp:
```diff
@@ -18,8 +18,28 @@
 
+    bool IsSdCardInserted();
+
+    Result OpenSdCardDetectionEventNotifier(SdCardDetectionEventNotifier *out);
```
I will handle this.

SciresM commented on libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_packet_transmitter.hpp:
```diff
@@ -0,0 +1,122 @@
+                return static_cast<size_t>(this->log_buffer_end - this->log_buffer_payload_current);
+            }
+
+            static constexpr size_t GetRequiredSizeToPushUleb128(u64 value) {
```
This seems fake/has messed up access controls. I'll refactor this myself after the detail -> impl changes are made.

SciresM commented on libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_packet_transmitter.hpp:
```diff
@@ -0,0 +1,122 @@
+            using LogPacketTransmitterBase::LogPacketTransmitterBase;
+
+            void PushLogSessionBegin() {
+                u8 dummy_value = 1;
```
This is actually the boolean value true.

SciresM commented on libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp:
```diff
@@ -0,0 +1,128 @@
+    static_assert(sizeof(LogPacketHeader) == 0x18);
+
+    enum LogDataChunkKey : u8 {
+        LogDataChunkKey_LogSessionBegin    = 0,  ///< Log session begin.
```
Atmosphere doesn't use doxygen, shouldn't have doxygen comments.

SciresM commented on libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp:
```diff
@@ -0,0 +1,128 @@
+    };
+
+    struct BinaryLogHeader {
+        static constexpr u32 Magic = util::FourCC<'h','p','h','p'>::Code;
```
Is this the correct endianness?

SciresM replied:
Given the code is conceptually "phph" (packet header, packet header), may well do to use reverse code with that encoding.

SciresM commented on libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp:
```diff
@@ -0,0 +1,128 @@
+        u64 thread_id;
+        u8 flags;
+        u8 pad;
+        diag::LogSeverity severity;
```
should be u8

SciresM commented on libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp:
```diff
@@ -0,0 +1,128 @@
+        u8 flags;
+        u8 pad;
+        diag::LogSeverity severity;
+        char verbosity;
```
should be u8

SciresM commented on libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp:
```diff
@@ -0,0 +1,128 @@
+            if (le) {
+                this->flags |= LogPacketFlags_LittleEndian;
+            }
+            else {
```
incorrect formatting for ams style

SciresM commented on libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp:
```diff
@@ -0,0 +1,128 @@
+            if (head) {
+                this->flags |= LogPacketFlags_Head;
+            }
+            else {
```
incorrect formatting for ams style

SciresM commented on libraries/libstratosphere/include/stratosphere/lm/detail/lm_detail_log_types.hpp:
```diff
@@ -0,0 +1,128 @@
+        }
+
+    };
+    static_assert(sizeof(LogPacketHeader) == 0x18);
```
static_assert(util::is_pod<LogPacketHeader>::value);

SciresM commented on libraries/libstratosphere/include/stratosphere/lm/lm_api.hpp:
```diff
@@ -0,0 +1,25 @@
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
```
gap

SciresM commented on libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:
```diff
@@ -0,0 +1,92 @@
+
+        constexpr const char *g_log_severity_color_table[LogSeverity_Count] = {
+            "\x1B[90m",        /* Dark gray for Trace. */
+            "",                /* No special color for Info. */
```
This is nullptr in Nintendo's implementation, I believe. Should we match?

SciresM commented on libraries/libstratosphere/include/stratosphere/diag/detail/diag_detail_log.hpp:
```diff
@@ -0,0 +1,37 @@
+
+    void LogImpl(const LogMetaData &log_metadata, const char *fmt, ...);
+    void VLogImpl(const LogMetaData &log_metadata, const char *fmt, std::va_list va_args);
+    void CallAllLogObserver(const LogMetaData &log_metadata, const LogBody &log_body);
```
It is, but it's internal/not exposed. Why are you exposing it?

SciresM commented on libraries/libstratosphere/include/stratosphere/diag/diag_log_observer.hpp:
```diff
@@ -0,0 +1,38 @@
+    void RegisterLogObserver(LogObserverHolder *observer_holder);
+    void UnregisterLogObserver(LogObserverHolder *observer_holder);
+
+    void ResetDefaultLogObserver();
```
They are, but it's internal/not exposed. Why are you exposing it?

SciresM commented on libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:
```diff
@@ -0,0 +1,92 @@
+        /* Get the color according to the log severity. */
+        if (log_metadata.log_severity < LogSeverity_Count) {
+            color_type = g_log_severity_color_table[log_metadata.log_severity];
+        }
```
line break after here.

SciresM commented on libraries/libstratosphere/include/stratosphere/diag/detail/diag_detail_string.hpp:
```diff
@@ -0,0 +1,24 @@
+
+namespace ams::diag::detail {
+
+    void PrintDebugString(const char *str, size_t len);
```
It does, but it's internal, and should probably not be exposed.

SciresM commented on libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:
```diff
@@ -0,0 +1,92 @@
+
+        if (!has_module_name && !color_type) {
+            /* Just send the text log. */
+            PrintDebugString(log_body.log_text, log_body.log_text_length);
```
return PrintDebugString

SciresM commented on libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:
```diff
@@ -0,0 +1,92 @@
+        }
+        auto has_module_name = false;
+        if (log_metadata.module_name != nullptr) {
+            has_module_name = strlen(log_metadata.module_name) > 1; /* Avoid '$' as module name. */
```
Comments go on their own lines.

SciresM commented on libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:
```diff
@@ -0,0 +1,92 @@
+        if (log_metadata.log_severity < LogSeverity_Count) {
+            color_type = g_log_severity_color_table[log_metadata.log_severity];
+        }
+        auto has_module_name = false;
```
...

const auto has_module_name = log_metadata.module_name != nullptr && std::strlen(log_metadata.module_name) > 1?

SciresM commented on libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:
```diff
@@ -0,0 +1,92 @@
+
+        size_t used_buffer_space = 0;
+        if (has_module_name && log_body.log_is_head) {
+            auto fmt_prefix = (color_type != nullptr) ? color_type : "";
```
const auto?

SciresM commented on libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:
```diff
@@ -0,0 +1,92 @@
+        if (has_module_name && log_body.log_is_head) {
+            auto fmt_prefix = (color_type != nullptr) ? color_type : "";
+
+            auto module_name = log_metadata.module_name;
```
what type is this? inappropriate auto usage.

SciresM commented on libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:
```diff
@@ -0,0 +1,92 @@
+            }
+
+            auto system_tick = os::GetSystemTick();
+            auto tick_ts = os::ConvertToTimeSpan(system_tick);
```
... const auto tick = os::ConvertToTimeSpan(os::GetSystemTick()); ?

SciresM commented on libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:
```diff
@@ -0,0 +1,92 @@
+            auto tick_ts = os::ConvertToTimeSpan(system_tick);
+            used_buffer_space += util::SNPrintf(g_default_log_observer_text_buffer, sizeof(g_default_log_observer_text_buffer), "%s%d:%02d:%02d.%03d [%-5s] ", fmt_prefix, tick_ts.GetHours(), tick_ts.GetMinutes(), tick_ts.GetSeconds(), tick_ts.GetMilliSeconds(), module_name);
+        }
+        else if (color_type) {
```
invalid else format

SciresM replied:
color_type != nullptr

SciresM commented on libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:
```diff
@@ -0,0 +1,92 @@
+        }
+
+        /* If we used a color before, reset it now. */
+        auto fmt_prefix = (color_type != nullptr) ? "\x1B[0m" : "";
```
Why is this a local variable, and not e.g. a constexpr variable which is ternary'd in the SNPrintf call?

SciresM commented on libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:
```diff
@@ -0,0 +1,92 @@
+            used_buffer_space += util::SNPrintf(g_default_log_observer_text_buffer, sizeof(g_default_log_observer_text_buffer), "%s", color_type);
+        }
+
+        auto left_buffer_space = sizeof(g_default_log_observer_text_buffer) - used_buffer_space;
```
...const auto max_buffer_size = sizeof(....) - used_buffer_space - (color_type != nullptr ? std::strlen(ResetColorType) - 1 : 0)

SciresM commented on libraries/libstratosphere/source/diag/detail/diag_detail_default_log_observer.cpp:
```diff
@@ -0,0 +1,92 @@
+            left_buffer_space -= strlen(color_type);
+        }
+
+        auto log_text_size = std::min(left_buffer_space, log_body.log_text_length);
```
const auto, specify that it's std::min<size_t>

SciresM commented on libraries/libstratosphere/source/diag/detail/diag_detail_log.cpp:
```diff
@@ -0,0 +1,128 @@
+            char buffer[40];
+        };
+
+        void PutCharacters(void *user_data, const char *chars, int chars_len) {
```
This needs comments, to not use auto everywhere, to use const where appropriate...

SciresM commented on libraries/libstratosphere/source/fs/fs_sd_card.cpp:
```diff
@@ -43,6 +43,22 @@ namespace ams::fs {
+        AMS_ABORT_UNLESS(this->notifier != nullptr);
+
+        ::Event event;
+        R_TRY(fsEventNotifierGetEventHandle(static_cast<FsEventNotifier *>(this->notifier), std::addressof(event), clear_mode == os::EventClearMode_AutoClear));
```
This needs fixes, which I'll do.

SciresM commented on libraries/libstratosphere/source/diag/diag_log_observer.cpp:
```diff
@@ -0,0 +1,89 @@
+ */
+#include <stratosphere.hpp>
+
+namespace ams::diag {
```
This and the other log observer file should really be a single translation unit. I'll handle the merge.

SciresM commented on libraries/libstratosphere/source/lm/detail/lm_detail_log_observer.cpp:
```diff
@@ -0,0 +1,100 @@
+    namespace {
+
+        os::SdkMutex g_log_observer_lock;
+        u8 g_packet_transmitter_buffer[0x400];
```
You only have a single buffer, N has a unit heap of 4, but I can handle this.

SciresM commented on libraries/libstratosphere/source/lm/detail/lm_detail_log_observer.cpp:
```diff
@@ -0,0 +1,100 @@
+
+    }
+
+    void LogManagerLogObserver(const diag::LogMetaData &log_metadata, const diag::LogBody &log_body, void *user_data) {
```
This (and other code) creates lots of locals that are non-const and never used again. Better off with immediates.

SciresM reviewed (changes requested):
Reviewed about half the files (no actual lm process code yet).

Mostly stylistic stuff. Do a refactor for the ::detail:: -> ::impl::, ::impl:: -> ::srv:: stuff mentioned before, then I'll do a pass over the lm process code.
